src/functions.jl

```diff
@@ -6,649 +6,646 @@
+                atan!(c, a, r, k)
+            end
+            return c
+        end
 
-with $a$ a `Taylor1` polynomial.
+        ## sinh and cosh ##
+        sinh(a::$T) = sinhcosh(a)[1]
+        cosh(a::$T) = sinhcosh(a)[2]
+        function sinhcosh(a::$T)
+            if $T == Taylor1
+                order = a.order
+                @inbounds s = $T( sinh(a[1]), order)
+                @inbounds c = $T( cosh(a[1]), order)
+            else
+                order = get_order()
+                @inbounds s = $T( sinh(a[1][1]), order)
+                @inbounds c = $T( cosh(a[1][1]), order)
+            end
+
+            @inbounds for k = 1:order
+                sinhcosh!(s, c, a, k)
+            end
+            return s, c
+        end
 
-Inputs are the `kcoef`-th coefficient, the vector of the expansion coefficients
-`ac` of `a`, and the already calculated expansion coefficients `coeffs` of `c`.
-"""
-function logHomogCoef{T<:Number}(kcoef::Int, ac::Array{T,1}, coeffs::Array{T,1})
-    kcoef == 0 && return log( ac[1] )
-    coefhomog = zero(T)
-    @inbounds for i = 1:kcoef-1
-        coefhomog += (kcoef-i) * ac[i+1] * coeffs[kcoef-i+1]
+        ## tanh ##
+        function tanh(a::$T)
+            if $T == Taylor1
+                order = a.order
+                @inbounds aux = tanh( a[1] )
+                c = $T( aux, order)
+                c2 = $T( aux^2, order)
+            else
+                order = get_order()
+                @inbounds aux = tanh( a[1][1] )
+                c = $T( aux, order)
+                c2 = $T( aux^2, order)
+            end
+
+            @inbounds for k = 1:order
+                tanh!(c, a, c2, k)
+            end
+            return c
+        end
     end
-    @inbounds coefhomog = (ac[kcoef+1] -coefhomog/kcoef) / ac[1]
-    coefhomog
 end
 
 
-### TRIGONOMETRIC FUNCTIONS ###
-
-## Sin ## 
-doc"""
-    sin(a)
-
-Return the Taylor expansion of $\sin(a)$, of order `a.order`, for
-`a::Taylor1` polynomial.
-
-For details on making the Taylor expansion, see
-[`TaylorSeries.sincosHomogCoef`](@ref).
-"""
-sin(a::Taylor1) = sincos(a)[1]
-
-## Cos ## 
-doc"""
-    cos(a)
-
-Return the Taylor expansion of $\cos(a)$, of order `a.order`,
-for `a::Taylor1` polynomial
-
-For details on making the Taylor expansion, see
-[`TaylorSeries.sincosHomogCoef`](@ref).
-"""
-cos(a::Taylor1) = sincos(a)[2]
-
-## Sin and Cos ## 
-function sincos(a::Taylor1)
-    @inbounds aux = sin( a[1] )
-    T = typeof(aux)
-    v = convert(Array{T,1}, a.coeffs)
-    sincoeffs = similar(v)
-    coscoeffs = similar(v)
-    @inbounds sincoeffs[1] = aux
-    @inbounds coscoeffs[1] = cos( a[1] )
-    @inbounds for k = 1:a.order
-        sincoeffs[k+1], coscoeffs[k+1] = sincosHomogCoef(k, v, sincoeffs, coscoeffs)
-    end
-    return Taylor1(sincoeffs, a.order), Taylor1(coscoeffs, a.order)
-end
+# Recursive functions (homogeneous coefficients)
+for T in (:Taylor1, :TaylorN)
+    @eval begin
+        function exp!(c::$T, a::$T, k::Int)
```

Comment:
Maybe it would be useful to use https://github.com/alsam/OffsetArrays.jl
to do away with all these annoying `+1` everywhere.